reports-backend/src/main/java/org/jboss/da/reports/impl/ReportsGeneratorImpl.java

```diff
@@ -81,35 +88,198 @@
+    }
+
+    private Optional<ArtifactReport> createScmReport(SCMLocator scml,
+            Set<ProductVersion> productVersions, Set<Long> productVersionIds) throws ScmException,
+            PomAnalysisException, CommunicationException {
+
         GAVDependencyTree dt = dependencyTreeGenerator.getDependencyTree(scml);
 
+        return createReport(dt, productVersions, productVersionIds);
+    }
+
+    private Optional<ArtifactReport> createReport(GAVDependencyTree dt,
+            Set<ProductVersion> productVersions, Set<Long> productVersionIds)
+            throws CommunicationException {
+
         VersionLookupResult result = versionFinder.lookupBuiltVersions(dt.getGav());
         ArtifactReport report = toArtifactReport(dt.getGav(), result);
 
         Set<GAVDependencyTree> nodesVisited = new HashSet<>();
         nodesVisited.add(dt);
         addDependencyReports(report, dt.getDependencies(), nodesVisited);
+        addArtifactsFromWhiteList(report, productVersions);
 
         return Optional.of(report);
     }
 
+    private void addArtifactsFromWhiteList(ArtifactReport report,
+            Set<ProductVersion> productVersions) {
+        Map<WhiteArtifact, Set<ProductVersion>> whiteArtifactsAndOrigin = getWhiteArtifacts(productVersions);
+        updateArtifact(report, whiteArtifactsAndOrigin);
+    }
+
+    private void updateArtifact(ArtifactReport ar,
+            Map<WhiteArtifact, Set<ProductVersion>> allArtifacts) {
+
+        Set<ArtifactReport> dependencies = ar.getDependencies();
+
+        // Process the current artifact
+
+        GAV currentGav = ar.getGav();
+
+        List<String> availableVersionsFromWhitelist = getAvailableWhitelistVersions(allArtifacts,
+                currentGav);
+        Optional<String> bestVersion = getBestMatchVersionFromWhitelist(ar.getBestMatchVersion(),
+                currentGav, availableVersionsFromWhitelist);
+
+        if (ar.getAvailableVersions() != null)
+            availableVersionsFromWhitelist.addAll(ar.getAvailableVersions());
+
+        // Update available versions
+        ar.setAvailableVersions(availableVersionsFromWhitelist);
+
+        // Update best match version
+        ar.setBestMatchVersion(bestVersion);
+
+        // Recurse or end
+        if (dependencies == null || dependencies.isEmpty()) {
+            return;
+        } else {
+            for (ArtifactReport depArtifact : dependencies) {
+                updateArtifact(depArtifact, allArtifacts);
+            }
+        }
+    }
+
+    private Optional<String> getBestMatchVersionFromWhitelist(Optional<String> currentBestVersion,
+            GAV currentGav, List<String> availableVersionsFromWhitelist) {
+        // Establish best match version
+        Optional<String> bestWhiteVersion = versionFinder.getBestMatchVersionFor(currentGav,
+                availableVersionsFromWhitelist);
+        if (bestWhiteVersion != null && bestWhiteVersion.isPresent()) {
```

Comment:
Ok, I think I've got it - I was running ReportsGeneratorImplTest and getting a null pointer because the mock versionFinder was probably being hit with a GAV it wasn't initialized for. The Arquillian tests never had a null pointer here - so not a bug.